### PR TITLE
feat: hide / unhide tokens

### PR DESCRIPTION
**Pull request by magiziz**
Fixes BX-1356, BX-1365

## What changed (plus any additional context for devs)

1. Context menu UI in the token list is now sorted by the right order. Also added keyboard shortcut for pin / unpin and hide / unhide.
2. Token details more menu is also sorted by the right order, but without shortcuts.
3. Added translations for hide, unhide and view on coingecko.
4. Added shorcuts to pin / unpin, hide /unhide and copy address + tracking (analytics)

## Screen recordings / screenshots

https://github.com/rainbow-me/browser-extension/assets/53529533/5b2ced74-cb8e-4931-a8f0-b93e5eca62ee

## What to test

1. In the token list right click on any token and you should see a overflow menu with option "Hide x token".
5. Hidden tokens should be hidden from token list, swap, send and bridge, but you should be able to unhide the token in token details page and token shouldn't be hidden when you do `cmd + K`.
6. For some tokens like testnet or custom network ones, the dropdown is usually hidden in the details page, but if you hide  testnet or custom network tokens the dropdown should appear to unhide the token.
7. Try to also pin a token, then hide it. It should also trigger unpin behind the scenes. You'll see it ones you unhide the token in the token details page, then go back to token list section in the homepage.

### CODE REVIEW

File: src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx
Comment (magiziz) [reply src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx:170]: Great idea 💡, let me do that 🙏

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (estebanmino) [inline src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx:437]: same here, isHidden could receive token as param and then inside of it you can use token.uniqueId

**Review by estebanmino — APPROVED** [review]
🪨

**Comment by magiziz** [comment]
For anyone who is reviewing this you might see that i don't use `uniqueId` to store hidden assets, but instead we store token address with chain id (`${asset.address}-${asset.chainId}`). This is because swap page has different `uniqueId`'s than token list page. Works for now, but might consider a follow up refactor PR once this is fixed in the backend.

File: src/entries/popup/hooks/useHiddenAssets.ts
Comment (greg-schrammel) [inline src/entries/popup/hooks/useHiddenAssets.ts:7]: why are these methods not in the store itself?

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (greg-schrammel) [inline src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx:298]: if AccentColorProvider wraps it all do we need the fragment?

File: src/entries/popup/hooks/useHiddenAssets.ts
Comment (magiziz) [reply src/entries/popup/hooks/useHiddenAssets.ts:7]: Yeah. We could do it from store itself, but there would be a lot of duplicate functions since we use the same `${asset.address}-${asset.chainId}` string to add and remove hidden assets + `isHidden` is used a lot so didn't want to duplicate that as well. 

Keeping this in a hook would just make things simpler. Lmk what you think 🙏

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (magiziz) [reply src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx]: Oohh great catch 👀. Fixed it in this [commit](https://github.com/rainbow-me/browser-extension/pull/1402/commits/85a314b3b5f1cb1c8a7c1be6cabe738f41ac7909).

File: src/entries/popup/hooks/useHiddenAssets.ts
Comment (greg-schrammel) [reply src/entries/popup/hooks/useHiddenAssets.ts:7]: I mean something like 
```ts
export const hiddenAssets = createStore<HiddenAssetState>(
  (set, get) => ({
    hiddenAssets: [],
    addHiddenAsset: (asset: { chainId: ChainId, address: Address }) => {
      const { hiddenAssets } = get();
      const uniqueId = `${asset.address}-${asset.chainId}`
      set({ hiddenAssets: [...hiddenAssets, uniqueId] });
    },
    isHidden: (asset) => {
      const { hiddenAssets } = get();
      const uniqueId = `${asset.address}-${asset.chainId}`
      return hiddenAssets.includes(uniqueId)
    }
```
so whats outside the store don't need to know about how it stores its data, just pass an asset and it handles the rest

Comment (greg-schrammel) [reply src/entries/popup/hooks/useHiddenAssets.ts:7]: and basically rename `useHiddenAssetStore` -> `useHiddenAssets` I think would work the same

**Review by greg-schrammel — APPROVED** [review]

File: src/entries/popup/hooks/useHiddenAssets.ts
Comment (magiziz) [reply src/entries/popup/hooks/useHiddenAssets.ts:7]: Ah ok. Yeah makes sense now, let me do that. Thanks 🙏

**Comment by DanielSinclair** [comment]
@magiziz Before you merge, going to send some tweaks to the context menu orders. Looks like a few regressions over time there.

File: static/json/languages/en_US.json
Comment (magiziz) [inline static/json/languages/en_US.json:1461]: Didnt want to use `view_on` since that's related to explorer's only

**Review by DanielSinclair — CHANGES_REQUESTED** [review]
cc @greg-schrammel Will need another review early next week after some bug fixes

File: lavamoat/build-webpack/policy.json
Comment (DanielSinclair) [inline lavamoat/build-webpack/policy.json:1]: Some strange policy changes here; might want to drop. Maybe an issue with your styling plugin?

File: src/entries/popup/pages/home/TokenDetails/KeyboardShortcutListener.tsx
Comment (DanielSinclair) [inline src/entries/popup/pages/home/TokenDetails/KeyboardShortcutListener.tsx:28]: These types would be different between Tokens (Home) and Token Details. Should create a subset type for token details shortcuts

File: src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx
Comment (DanielSinclair) [inline src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx:218]: We have a `ContextMenuSeparator` type for this behavior; should use that

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (DanielSinclair) [inline src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx:368]: Also have a `<DropdownMenuSeperator>` component for this

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (DanielSinclair) [inline src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx:318]: Looks like the shortcut hint UI is missing here 
Screenshot 2024-03-23 at 3 29 49 AM

File: src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx
Comment (DanielSinclair) [inline src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx:143]: Have a look at how we do this for the tokens screen (home) with the `useTokensShortcuts` hook. I think we want to continue with consistency of that approach rather than a wrapper
https://github.com/rainbow-me/browser-extension/blob/764e5153b9c746cb7cf20f79ee46d6e103df3ec8/src/entries/popup/hooks/useTokensShortcuts.ts#L23

File: static/json/languages/en_US.json
Comment (DanielSinclair) [inline static/json/languages/en_US.json:1472]: Check out the `hide_balances_toast` example; I think we want to bump this out of this section and have a separate toast key for these actions. Try this for a string: `Your %{name} is hidden`. Would also need a toast for unhide and pin/unpin I think. I realized after my prior comment that we do want a toast on the Token Details page for those, but don't need one on Tokens (because of the visual list shift)

File: src/entries/popup/hooks/useHiddenAssets.ts
Comment (DanielSinclair) [inline src/entries/popup/hooks/useHiddenAssets.ts:1]: Don't think we need this file. Generally we use hooks for read-only state (i.e. the full list of wallets, or memoized versions of that list), whereas these are duplicate mutations already represented in `useHiddenAssetStore`. Should probably use that directly, and then perform the `isHidden` check inline in the components based on the results of `hiddenAssets`

File: src/core/state/hiddenAssets/hiddenAssets.ts
Comment (DanielSinclair) [inline src/core/state/hiddenAssets/hiddenAssets.ts:17]: `hiddenAssets` should be `hiddenAssetsStore` here

Comment (DanielSinclair) [reply src/core/state/hiddenAssets/hiddenAssets.ts:17]: If we needed a hidden assets hook at some point (like to display a list of just the hidden assets) we wouldn't want the name collision

File: src/core/state/hiddenAssets/hiddenAssets.ts
Comment (derHowie) [inline src/core/state/hiddenAssets/hiddenAssets.ts:24]: Is there a particular reason you're building this yourself? Now that https://github.com/rainbow-me/browser-extension/commit/b1e3f5edc7d328978f07338eab49025b372c5f28 is merged, would it make more sense to grab this from the asset itself? Haven't fully thought this through, but would be curious to hear your reasoning.

File: src/entries/popup/pages/home/Tokens.tsx
Comment (derHowie) [inline src/entries/popup/pages/home/Tokens.tsx:173]: The naming in this file is getting kind of wonky. We are creating `combinedFilteredAssets` then using that value to create `filteredAssets`.

Maybe we could describe the filter happening in the first value so it's easier to think about later. Something like `unhiddenAssets` or `assetsToDisplay` or `assetsToShow`?

File: src/entries/popup/pages/send/index.tsx
Comment (derHowie) [inline src/entries/popup/pages/send/index.tsx:137]: Whatever name you choose for the above comment, i'd mirror that here since we're doing the same operation.

File: src/entries/popup/pages/home/Tokens.tsx
Comment (derHowie) [inline src/entries/popup/pages/home/Tokens.tsx:305]: My guess is that you're passing this all the way down to the context menus so that if we use a shortcut, we can simulate a click on the main container to close the context menu?

Do you think there'd be negative performance implications in just using this hook in the context menu itself rather than prop drilling it down two levels?

**Review by derHowie — COMMENTED** [review]
Looks good, just a couple questions.

File: src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx
Comment (derHowie) [inline src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx:146]: Do you think this functionality could live within `useTokenShortcuts` with the other token shortcut handlers?

Comment (magiziz) [reply src/entries/popup/pages/home/TokenDetails/TokenContextMenu.tsx:146]: Yeah was thinking of doing that too, but realised we already have it as a global scope [here](https://github.com/rainbow-me/browser-extension/blob/master/src/entries/popup/pages/home/Tokens.tsx#L229) in the homepage. I could take the hook and listen to it again, but we would face duplications which would cause double tracking.

File: src/entries/popup/pages/home/Tokens.tsx
Comment (magiziz) [reply src/entries/popup/pages/home/Tokens.tsx]: Good point. Fixing rn

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (magiziz) [reply src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx]: We're going to fix this in another ticket. Seems like there is a bug with `<DropdownItem />`

**Review by derHowie — APPROVED** [review]
LGTM, thanks for the changes

File: src/entries/popup/hooks/useTokensShortcuts.ts
Comment (greg-schrammel) [inline src/entries/popup/hooks/useTokensShortcuts.ts:97]: are these simulateClick on the container to close the context menu?

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (greg-schrammel) [inline src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx:391]: we have `goToNewTab` function somewhere, not sure why 
@estebanmino is there a reason to use `chrome.tabs.create` instead of `window.open`?

File: src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx
Comment (greg-schrammel) [inline src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx:320]: if your not passing this function to a memoized component it don't need keep it's ref stable (`useCallback`)

Comment (greg-schrammel) [reply src/entries/popup/pages/home/TokenDetails/TokenDetails.tsx:320]: because like it is recreating the function anyway, just going thru more effort to check the cache and return the previous if deps didn't change
I mean `useCallback` is super cheap but not free

not a change request, just a note

File: src/entries/popup/hooks/useTokensShortcuts.ts
Comment (magiziz) [reply src/entries/popup/hooks/useTokensShortcuts.ts:97]: Yep

**Review by greg-schrammel — APPROVED** [review]

**Review by DanielSinclair — APPROVED** [review]

**Pull request merged**
